### PR TITLE
Add scripts about CSP policies notification feature and SameSite cookie protection

### PR DESCRIPTION
**Pull request by righettod**
Hello,

_First script:_
The goal of the proposed script "**detect_samesite_protection.js**" is to detect if the site use the protection bring by the "SameSite" cookie attribute.

Knowing that point is interesting because the goal of this attribute is to mitigate CSRF attack.

_Second script:_
The goal of the proposed script "**detect_csp_notif_and_reportonly.js**" is to detect the following points about CSP policies defined by a site:

- If policies send notifications about violations (some kind of monitoring by the site owner),
- If policies behave into "report-only" mode (no blocking, only report violations).

These informations are interesting from an attacker/researcher point of view because it indicates to him that:

- His input validation probing tentatives will be potentially quickly detected (depending on the monitoring level by site owner),
- The CSP policies in place will not block the exploitation if a vulnerability is found into input validation,
- Perhaps the endpoint receiving notifications is vulnerable to some injection.

Thanks in advance for your feedback.

Regards.

### CODE REVIEW

File: passive/detect_csp_notif_and_reportonly.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+These informations are interesting from an attacker/researcher point of view because it indicates to him:
+1) His input validation probing tentatives will be potentially quickly detected (depending on the monitoring level by site owner),
+2) The CSP policies in place will not block the exploitation if a vulnerability is found into input validation,
+3) Perhaps the endpoint receiving notifications is vulnerable to some injection.
+
+Links:
+- http://content-security-policy.com
+- http://www.html5rocks.com/en/tutorials/security/content-security-policy
+- http://www.html5rocks.com/en/tutorials/security/content-security-policy/#reporting
+*/
```
Comment (thc202) [inline passive/detect_csp_notif_and_reportonly.js]: The author entry was removed, was that on purpose?

File: passive/detect_csp_notif_and_reportonly.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	// raiseAlert(risk, int confidence, String name, String description, String uri, 
+	//		String param, String attack, String otherInfo, String solution, String evidence, 
+	//		int cweId, int wascId, HttpMessage msg)
+	// risk: 0: info, 1: low, 2: medium, 3: high
+	// confidence: 0: falsePositive, 1: low, 2: medium, 3: high, 4: confirmed
+
+	//Common variables
+	var cweId = 200; 
+	var wascId = 13;
+	var url = msg.getRequestHeader().getURI().toString();
```
Comment (thc202) [inline passive/detect_csp_notif_and_reportonly.js]: There's a tab at the end of the line.

File: passive/detect_csp_notif_and_reportonly.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+					//In CSP specification, policies are separated by ';'
+					var startPosition = cspPolicies.search(cspReportInstruction);
+					var tmp = cspPolicies.substring(startPosition);
+					var endPosition = tmp.indexOf(";");
+					if(endPosition != -1){
+						var reportUrl = tmp.substring(0, endPosition);	
+					}else{
+						var reportUrl = tmp;
+					}
+					reportUrl = reportUrl.replace(cspReportInstruction,"").trim();
```
Comment (thc202) [inline passive/detect_csp_notif_and_reportonly.js]: There are some tabs at the end of the line.

File: passive/detect_csp_notif_and_reportonly.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+					if(endPosition != -1){
+						var reportUrl = tmp.substring(0, endPosition);	
+					}else{
+						var reportUrl = tmp;
+					}
+					reportUrl = reportUrl.replace(cspReportInstruction,"").trim();					
+					//Raise info alert
+					var cspWorkingMode = (headerName.toLowerCase().indexOf("-report-only") == -1) ? "BLOCKING" : "REPORTING";
+					var description = "The current site CSP policies defined by HTTP response header '" + headerName + "' (behaving in " + cspWorkingMode + " mode) report violation to '" + reportUrl + "'.";
+					var infoLinkRef = "https://developer.mozilla.org/en-US/docs/Web/Security/CSP/Using_CSP_violation_reports";
```
Comment (thc202) [inline passive/detect_csp_notif_and_reportonly.js]: There's a tab at the end of the line.

File: passive/detect_csp_notif_and_reportonly.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+						var reportUrl = tmp.substring(0, endPosition);	
+					}else{
+						var reportUrl = tmp;
+					}
+					reportUrl = reportUrl.replace(cspReportInstruction,"").trim();					
+					//Raise info alert
+					var cspWorkingMode = (headerName.toLowerCase().indexOf("-report-only") == -1) ? "BLOCKING" : "REPORTING";
+					var description = "The current site CSP policies defined by HTTP response header '" + headerName + "' (behaving in " + cspWorkingMode + " mode) report violation to '" + reportUrl + "'.";
+					var infoLinkRef = "https://developer.mozilla.org/en-US/docs/Web/Security/CSP/Using_CSP_violation_reports";	
+					var solution = "Site owner will be notified at each policies violations, so, start by analyzing if a real monitoring of the notifications is in place before to use fuzzing or to be more aggressive.";
```
Comment (thc202) [inline passive/detect_csp_notif_and_reportonly.js]: There are some tabs at the end of the line.

File: passive/detect_samesite_protection.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
+/*
+Script to detect if the site use the protection bring by the "SameSite" cookie attribute.
+
+Knowing that point is interesting because the goal of this attribute is to mitigate CSRF attack.
+
+Links:
+- https://chloe.re/2016/04/13/goodbye-csrf-samesite-to-the-rescue
+- https://tools.ietf.org/html/draft-west-first-party-cookies
+- https://www.chromestatus.com/feature/4672634709082112
+*/
```
Comment (thc202) [inline passive/detect_samesite_protection.js]: The author entry was removed, was that on purpose?

File: passive/detect_samesite_protection.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	// raiseAlert(risk, int confidence, String name, String description, String uri, 
+	//		String param, String attack, String otherInfo, String solution, String evidence, 
+	//		int cweId, int wascId, HttpMessage msg)
+	// risk: 0: info, 1: low, 2: medium, 3: high
+	// confidence: 0: falsePositive, 1: low, 2: medium, 3: high, 4: confirmed
+
+	//Common variables
+	var cweId = 352; 
+	var wascId = 9;
+	var url = msg.getRequestHeader().getURI().toString();
```
Comment (thc202) [inline passive/detect_samesite_protection.js]: There's a tab at the end of the line.

File: passive/detect_samesite_protection.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+				//Inspect each attribute in order to avoid false-positive spot
+				//by simply searching "samesite=" on the whole cookie header value...
+				for(var k = 0 ; k < cookieAttributes.length ; k++){
+					var parts = cookieAttributes[k].split("=");
+					if(parts[0].trim().toLowerCase() == cookieSameSiteAttributeNameLower){
+						//Raise info alert
+						var sameSiteAttrValue = parts[1].trim();
+						var cookieName = cookieAttributes[0].split("=")[0].trim();
+						var description = "The current site use the 'SameSite' cookie attribute protection on cookie named '" + cookieName + "', value is set to '" + sameSiteAttrValue + "' protection level.";
+						var infoLinkRef = "https://tools.ietf.org/html/draft-west-first-party-cookies\nhttps://chloe.re/2016/04/13/goodbye-csrf-samesite-to-the-rescue";
```
Comment (thc202) [inline passive/detect_samesite_protection.js]: There are some tabs at the end of the line (same for the following to lines).

**Comment by thc202** [comment]
Just minor comments mostly, scripts work fine in Rhino and Nashorn. Thanks!

File: passive/detect_csp_notif_and_reportonly.js (hunk 1 above)
Comment (righettod) [inline passive/detect_csp_notif_and_reportonly.js]: Error from my side, now fixed.

File: passive/detect_csp_notif_and_reportonly.js (hunk 2 above)
Comment (righettod) [inline passive/detect_csp_notif_and_reportonly.js]: Fixed

File: passive/detect_csp_notif_and_reportonly.js (hunk 3 above)
Comment (righettod) [inline passive/detect_csp_notif_and_reportonly.js]: Fixed

File: passive/detect_csp_notif_and_reportonly.js (hunk 4 above)
Comment (righettod) [inline passive/detect_csp_notif_and_reportonly.js]: Fixed

File: passive/detect_csp_notif_and_reportonly.js (hunk 5 above)
Comment (righettod) [inline passive/detect_csp_notif_and_reportonly.js]: Fixed

File: passive/detect_samesite_protection.js (hunk 6 above)
Comment (righettod) [inline passive/detect_samesite_protection.js]: Error from my side, now fixed.

File: passive/detect_samesite_protection.js (hunk 7 above)
Comment (righettod) [inline passive/detect_samesite_protection.js]: Fixed

**Comment by righettod** [comment]
All your remarks has been taken in account.
The thing that I cannot achieve is merging the 2 commits of the pull request 37 into a single one.
If you can help me on this point it will be really cool :)

File: passive/detect_samesite_protection.js (hunk 8 above)
Comment (righettod) [inline passive/detect_samesite_protection.js]: Fixed

File: passive/detect_csp_notif_and_reportonly.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	if(cspPolicies.indexOf(cspReportInstruction) != -1){
+		var startPosition = cspPolicies.search(cspReportInstruction);
+		var tmp = cspPolicies.substring(startPosition);
+		var endPosition = tmp.indexOf(";");
+		if(endPosition != -1){
+			var reportUrl = tmp.substring(0, endPosition);	
+		}else{
+			var reportUrl = tmp;
+		}
+		return reportUrl.replace(cspReportInstruction,"").trim();
```
Comment (thc202) [inline passive/detect_csp_notif_and_reportonly.js]: There are some tabs at the end of the line.

**Comment by thc202** [comment]
The commit message could just be "Add scripts about CSP policies notification feature and SameSite cookie protection".

(You can use `--amend` to address both comments.)

**Comment by righettod** [comment]
I have made a mistake during my rebase and push that have implied that I have to restart from zero (I have squashed a commit that is not mine).
I must close this pull request because it not linked with any source repository anymore...
I will create a new one, sorry.

**Comment by thc202** [comment]
No problem :)

**Pull request closed without merging**